File: interop/reference-ts/src/canonical.ts
/** Canonical JSON serialization and content-hash projection, mirroring
 * atmanatic_research/canonical.py. Object keys are sorted, separators are
 * compact, and non-finite numbers are rejected. `content_hash` and
 * `signature` are excluded from the hashed projection.
 */

import { createHash } from "node:crypto";

const EXCLUDED_HASH_FIELDS = new Set(["content_hash", "signature"]);

export class CanonicalizationError extends Error {}

function checkFinite(value: unknown): void {
  if (typeof value === "number" && !Number.isFinite(value)) {
    throw new CanonicalizationError(
      "canonical JSON does not permit NaN or infinite numbers"
    );
  }
  if (Array.isArray(value)) {
    for (const item of value) checkFinite(item);
  } else if (value !== null && typeof value === "object") {
    for (const item of Object.values(value as Record<string, unknown>)) {
      checkFinite(item);
    }
  }
}

/** Serialize `value` with sorted object keys and compact separators, matching
 * the Python reference's `json.dumps(value, sort_keys=True, separators=(",", ":"))`.
 */
function canonicalStringify(value: unknown): string {
  if (value === null || typeof value === "boolean" || typeof value === "number") {
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(canonicalStringify).join(",")}]`;
  }
  if (typeof value === "object") {
    const record = value as Record<string, unknown>;
    const keys = Object.keys(record).sort();
    const entries = keys.map(
      (key) => `${JSON.stringify(key)}:${canonicalStringify(record[key])}`
    );
    return `{${entries.join(",")}}`;
  }
  throw new CanonicalizationError(`cannot canonicalize value of type ${typeof value}`);
}

export function canonicalJsonBytes(value: unknown): Buffer {
  checkFinite(value);
  return Buffer.from(canonicalStringify(value), "utf-8");
}

export function projectForHash(
  record: Record<string, unknown>
): Record<string, unknown> {
  if (record === null || typeof record !== "object") {
    throw new CanonicalizationError("record must be an object");
  }
  const projected: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(record)) {
    if (!EXCLUDED_HASH_FIELDS.has(key)) projected[key] = value;
  }
  return projected;
}

export function computeContentHash(record: Record<string, unknown>): string {
  const projected = projectForHash(record);
  return createHash("sha256").update(canonicalJsonBytes(projected)).digest("hex");
}

export function verifyContentHash(record: Record<string, unknown>): boolean {
  const declared = record?.content_hash;
  if (typeof declared !== "string") return false;
  return declared.toLowerCase() === computeContentHash(record).toLowerCase();
}

File: atmanatic_research/canonical.py
"""Canonical JSON serialization and content-hash projection for artifacts.

This freezes the Protocol 0.1 canonicalization scheme described in
ATMANATIC_PROTOCOL_0.1_DRAFT.md section 5.2: sorted object keys, compact
separators, UTF-8 bytes, and no non-finite numbers. `content_hash` and
`signature` fields are excluded from the hashed projection so a record can
declare its own hash without creating a circular dependency.
"""

from __future__ import annotations

import hashlib
import json
import math
from typing import Any

EXCLUDED_HASH_FIELDS = frozenset({"content_hash", "signature"})


class CanonicalizationError(ValueError):
    """Raised when a value cannot be canonicalized deterministically."""


def _check_finite(value: Any) -> None:
    if isinstance(value, float) and not math.isfinite(value):
        raise CanonicalizationError("canonical JSON does not permit NaN or infinite numbers")
    if isinstance(value, dict):
        for item in value.values():
            _check_finite(item)
    elif isinstance(value, list):
        for item in value:
            _check_finite(item)


def canonical_json_bytes(value: Any) -> bytes:
    """Return the canonical UTF-8 JSON bytes for `value`.

    Object keys are sorted, separators are compact, and non-finite floats are
    rejected. Callers MUST pass only JSON-compatible values.
    """
    _check_finite(value)
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode(
        "utf-8"
    )


def project_for_hash(record: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of `record` with hash and signature fields removed."""
    if not isinstance(record, dict):
        raise CanonicalizationError("record must be an object")
    return {key: value for key, value in record.items() if key not in EXCLUDED_HASH_FIELDS}


def compute_content_hash(record: dict[str, Any]) -> str:
    """Compute the SHA-256 content hash of `record` under the canonical projection."""
    projected = project_for_hash(record)
    return hashlib.sha256(canonical_json_bytes(projected)).hexdigest()


def verify_content_hash(record: dict[str, Any]) -> bool:
    """Return whether `record["content_hash"]` matches its recomputed canonical hash."""
    declared = record.get("content_hash") if isinstance(record, dict) else None
    if not isinstance(declared, str):
        return False
    return declared.lower() == compute_content_hash(record).lower()
